source/extensions/filters/network/zookeeper_proxy/zookeeper_decoder.cc

```diff
@@ -0,0 +1,199 @@
+#include "extensions/filters/network/zookeeper_proxy/zookeeper_decoder.h"
+
+#include "extensions/filters/network/zookeeper_proxy/zookeeper_utils.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ZooKeeperProxy {
+
+void DecoderImpl::decode(Buffer::Instance& data, uint64_t& offset) {
+  ENVOY_LOG(trace, "zookeeper_proxy: decoding {} bytes at offset {}", data.length(), offset);
+
+  // Check message length.
+  int32_t len;
+  BufferHelper::peekInt32(data, offset, len);
+  if (len < 8) {
+    callbacks_.onDecodeError();
+    return;
+  }
+
+  int32_t xid;
+  BufferHelper::peekInt32(data, offset, xid);
+
+  // "Special" requests.
+  switch (xid) {
+  case CONNECT_XID:
+    parseConnect(data, offset, len);
+    return;
+  case PING_XID:
+    callbacks_.onPing();
+    return;
+  case AUTH_XID:
+    parseAuthRequest(data, offset, len);
+    return;
+  default:
+    break;
+  }
+
+  // "Regular" requests.
+  int32_t opcode;
+  BufferHelper::peekInt32(data, offset, opcode);
+  switch (opcode) {
+  case enumToInt(Opcodes::GETDATA):
+    parseGetDataRequest(data, offset, len);
+    break;
+  case enumToInt(Opcodes::CREATE):
+    parseCreateRequest(data, offset, len, false);
+    break;
+  case enumToInt(Opcodes::CREATE2):
+    parseCreateRequest(data, offset, len, true);
+    break;
+  case enumToInt(Opcodes::SETDATA):
+    parseSetRequest(data, offset, len);
+    break;
+  case enumToInt(Opcodes::GETCHILDREN):
+    parseGetChildrenRequest(data, offset, len, false);
+    break;
+  case enumToInt(Opcodes::GETCHILDREN2):
+    parseGetChildrenRequest(data, offset, len, true);
+    break;
+  default:
+    break;
+  }
+}
+
+void DecoderImpl::parseConnect(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  bool readonly{};
+
+  if (len < 28) {
+    return;
+  }
+
+  // Read password - skip zxid, timeout, and session id.
+  std::string passwd;
+  offset += 20;
+  BufferHelper::peekString(data, offset, passwd);
+
+  // Read readonly flag, if it's there.
+  BufferHelper::peekBool(data, offset, readonly);
+
+  callbacks_.onConnect(readonly);
+}
+
+void DecoderImpl::parseAuthRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  if (len < 20) {
+    return;
```

Comment:
They are not really errors, it's possible for the decoder to get up to this point by accident.. So silently returning is good.

This is just here to avoid even trying to extract an auth request, if the minimum required bytes are not there. Though, to your point from below the message lengths could be done more consistently...